include/pmacc/memory/buffers/Exchange.hpp

```diff
@@ -80,19 +289,32 @@ namespace pmacc
          *
          * The buffer can point to device or host memory.
          */
-        virtual Buffer<TYPE, DIM>* getCommunicationBuffer() = 0;
-
-        virtual bool hasDeviceDoubleBuffer() = 0;
+        Buffer<TYPE, DIM>* getCommunicationBuffer()
+        {
+            if(Environment<>::get().isMpiDirectEnabled())
+            {
+                if(hasDeviceDoubleBuffer())
+                    return &(getDeviceDoubleBuffer());
+                else
+                    return &(getDeviceBuffer());
+            }
 
-        virtual DeviceBuffer<TYPE, DIM>& getDeviceDoubleBuffer() = 0;
+            return &(getHostBuffer());
+        }
 
     protected:
-        Exchange(uint32_t extype, uint32_t tag) : exchange(extype), communicationTag(tag)
-        {
-        }
+        /** host double buffer of the exchange data
+         *
+         * Is always a nullptr if MPI direct is used
+         */
+        std::unique_ptr<HostBuffer<TYPE, DIM>> hostBuffer;
+
+        //! This buffer is a vector which is used as message buffer for faster memcopy
+        std::unique_ptr<DeviceBuffer<TYPE, DIM>> deviceDoubleBuffer;
+        std::unique_ptr<DeviceBuffer<TYPE, DIM>> deviceBuffer;
 
-        uint32_t exchange;
-        uint32_t communicationTag;
+        uint32_t exchange = std::numeric_limits<uint32_t>::max();
```

Comment:
I added the default so that everyone can see that this will definitively be initialized with something valid.
Any time where I see a member without a default I am asking myself if the value is set in the constructor or not.
I always need to find all constructors to be sure that they will have a valid value.  By giving a default here I know now even without a lookup that it has a valid value even if it is not part of the constructor.